File: 541-reverse-string-ii/reverse-string-ii.js
/**
 * @param {string} s
 * @param {number} k
 * @return {string}
 */
var reverseStr = function (s, k) {
    let arr = s.split('');

    for (let i = 0; i < arr.length; i += 2 * k) {
        let left = i;
        let remaining = arr.length - left;

        if (remaining < k) {
            reverse(arr, left, arr.length - 1);
        } else {
            reverse(arr, left, left + k - 1);
        }
    }

    return arr.join('');
};

const reverse = (arr, left, right) => {
    while (left < right) {
        const temp = arr[left];
        arr[left] = arr[right];
        arr[right] = temp;
        left++;
        right--;
    }
}

File: 541-reverse-string-ii/reverse-string-ii.java
class Solution {
    public String reverseStr(String s, int k) {
        char[] arr = s.toCharArray();
        int n = arr.length;

        for (int i = 0; i < n; i += 2 * k) {
            int left = i;
            int remaining = n - left;

            if (remaining < k) {
                reverse(arr, left, n - 1);
            } else {
                reverse(arr, left, left + k - 1);
            }
        }

        return new String(arr);
    }

    private void reverse(char[] arr, int left, int right) {
        while (left < right) {
            char temp = arr[left];
            arr[left] = arr[right];
            arr[right] = temp;
            left++;
            right--;
        }
    }
}
